pkg/istio/authorizationpolicy/controller.go

```diff
@@ -0,0 +1,549 @@
+	}
+	// dealing with service resource
+
+	if obj, ok := (castItem.Resource).(*corev1.Service); ok {
+
+		var convertedCR model.Config
+
+		if castItem.Operation == model.EventAdd {
+			// creation of service will check if istio annotation is set to true, will result in creation of authz policy
+			// authz policy creation logic
+			// check if current service has istio authz annotation set
+			if _, ok = obj.Annotations[authzEnabledAnnotation]; ok {
+				if obj.Annotations[authzEnabledAnnotation] == authzEnabled {
+					// here should check if authz policy is existing in the cluster
+					log.Infof("istio authz annotation for service %s is set to true\n", obj.Name)
+					// form the authorization policy config and send create sign to the queue
+					athenzDomainRaw, exists, err := c.adIndexInformer.GetIndexer().GetByKey(athenz.NamespaceToDomain(obj.Namespace))
+					if err != nil {
+						return err
+					}
+
+					if !exists {
+						// TODO, add the non existing athenz domain to the istio custom resource
+						// processing controller to delete them
+						return fmt.Errorf("athenz domain %v does not exist in cache", obj.Namespace)
+					}
+
+					athenzDomain, ok := athenzDomainRaw.(*adv1.AthenzDomain)
+					if !ok {
+						return errors.New("athenz domain cast failed")
+					}
+
+					signedDomain := athenzDomain.Spec.SignedDomain
+					labels := obj.GetLabels()
+					domainRBAC := m.ConvertAthenzPoliciesIntoRbacModel(signedDomain.Domain, &c.adIndexInformer)
+					convertedCR = c.convertAthenzModelIntoIstioAuthzPolicy(domainRBAC, obj.Namespace, obj.Name, labels["svc"])
+					log.Infoln("Creating Authz Policy ... ")
+					revision, err := c.configStoreCache.Create(convertedCR)
+					if err != nil {
+						log.Errorln("error creating authz policy: ", err.Error())
+						return err
+					}
+					log.Infoln("Created revision number is: ", revision)
+				} else {
+					// case when service has authz flag switch from true to false, authz policy with the same name present
+					if res := c.configStoreCache.Get(schemas.AuthorizationPolicy.Type, obj.Name, obj.Namespace); res != nil {
+						log.Infoln("Deleting Authz Policy ... ")
+						c.configStoreCache.Delete(schemas.AuthorizationPolicy.Type, obj.Name, obj.Namespace)
+					}
+				}
+			} else {
+				// case when service has authzEnabledAnnotation removed, and authz policy with the same name present
+				if res := c.configStoreCache.Get(schemas.AuthorizationPolicy.Type, obj.Name, obj.Namespace); res != nil {
+					log.Infoln("Deleting Authz Policy ... ")
+					c.configStoreCache.Delete(schemas.AuthorizationPolicy.Type, obj.Name, obj.Namespace)
+				}
+			}
+		} else if castItem.Operation == model.EventUpdate {
+			if _, ok = obj.Annotations[authzEnabledAnnotation]; ok {
+				if obj.Annotations[authzEnabledAnnotation] == authzEnabled {
+					if res := c.configStoreCache.Get(schemas.AuthorizationPolicy.Type, obj.Name, obj.Namespace); res == nil {
```

Comment:
Can we combine this logic with the create operation? We could make this a common function.

Reply:
created a common function for create